File: Python/DynamicStack.py
class DynamicStack:
    def __init__(self):
        self.capacity = 2
        self.stack = [None]*self.capacity
        self.top = 0

    def push(self,data):
        if self.size() == self.capacity:
            self._expand()
        self.stack[self.top] = data
        self.top+=1

    def _expand(self):
        self.capacity *= 2
        newStack = [None]*self.capacity
        for i in range(self.size()):
            newStack[i] = self.stack[i]
        self.stack = newStack

    def pop(self):
        if self.is_empty():
            print("Stack is Emmpty")
            return None
        self.top-=1
        data = self.stack[self.top]
        self.stack[self.top] = None
        return data

    def peek(self):
        if self.is_empty():
            print("Stack is Empty")
            return None
        return self.stack[self.top-1]

    def size(self):
        return self.top
    
    def is_empty(self):
        return self.top <= 0
    
    def show(self):
        print("Stack (top to bottom):", [self.stack[i] for i in range(self.top - 1, -1, -1)])


nums = DynamicStack()    
print("Empty ?", nums.is_empty()) 
nums.push(4)
nums.push(10)
nums.push(72)

print("Top Element is =", nums.peek())

nums.push(13)
nums.push(7)

print("Top Element is =", nums.peek())

nums.pop()

print("Top Element is =", nums.peek())
nums.show()   


        

File: Java/Stack/DynamicStack.java
package Java.Stack;

public class DynamicStack {
    
    int capacity = 2;
    int stack[] = new int[capacity];
    int top=0;

    public void push(int data)
    {
        if(size()==capacity)
        expand();
       stack[top] = data;
       top++;
        
    }
    private void expand()
    {
        int length = size();
        int newStack[] = new int[capacity*2];
        System.arraycopy(stack,0,newStack,0,length);
        stack = newStack;
        capacity *= 2;
    }

    public int pop()
    {

        int data =0;
        if(isEmpty())
        {
            System.out.println("Stack is Empty");
        }
        else
        {
        
        top--;
        data = stack[top];
        stack[top]=0;
        shrink();
        }
        return data;
        
    }

    private void shrink()
    {
        int length = size();
        if(length <= capacity/4 && capacity > 2)
        {
        capacity = capacity/2;

        int newStack[] = new int[capacity];
         System.arraycopy(stack,0,newStack,0,length);
//           public static void arraycopy(Object source_arr, int sourcePos,
//                             Object dest_arr, int destPos, int len)
// Parameters : 
// source_arr : array to be copied from
// sourcePos : starting position in source array from where to copy
// dest_arr : array to be copied in
// destPos : starting position in destination array, where to copy in
// len : total no. of components to be copied. 

        stack = newStack;
 
        }

    }
    
    public int peek()
    {
        int data ;
        data = stack[top-1];
        return data;
    }

    public int size()
    {
        return top;
    }

    public boolean isEmpty()
    {
        return top<=0;
    }
    public void show()
    {
        for(int n: stack)
        {
                System.out.println(n + " ");
        }
    }
}
